### PR TITLE
feat: add snapshot table in gitextractor plugin

### PR DESCRIPTION
**Pull request by xgdyp**
# Summary

I added a table into snapshot for calculating some metrics like line age.

### Does this close any open issues?
related to #1771 

### Screenshots
Include any relevant screenshots here.

### Other Information
Any other information that is important to this PR.

### CODE REVIEW

**Comment by xgdyp** [comment]
Using this repo_snapshot table, we can make a query like 
```
SELECT weekday(commits.committed_date) AS wd,
        count(*)
FROM repo_snapshot
JOIN commits
    ON commits.sha=repo_snapshot.commit_sha
WHERE repo_id LIKE 'github:GithubRepo:1:491450511'
GROUP BY  wd
ORDER BY  wd ;
```
to get in which day,  the code has highest chance to stay in repository.
In lake, the result is  Wednesday.
image
we can also use a query like
```
SELECT file_path,
        avg(timestampdiff(day,
        commits.committed_date,
        now())) AS line_age
FROM repo_snapshot
JOIN commits
    ON repo_snapshot.commit_sha = commits.sha
WHERE repo_id LIKE 'github:GithubRepo:1:491450511'
GROUP BY  file_path
ORDER BY  line_age desc;
```
to get the answer of which file has the longest average line age
and the top 10 results are :
image

**Comment by xgdyp** [comment]
Hello @klesh, I add a new table named snapshot to repo_snapshot. using this table we can archieve some metrics.
I also update grafana gitextractor dashboard too. Like:
image
If this is OK, I will create a new PR to submit the dashboard change.

File: plugins/gitextractor/models/migrationscripts/20220927_add_snapshot.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+package migrationscripts
+
+import (
+	"context"
+	"github.com/apache/incubator-devlake/errors"
+	"github.com/apache/incubator-devlake/models/common"
+	"gorm.io/gorm"
+)
+
+type Snapshot struct {
```
Comment (klesh) [inline plugins/gitextractor/models/migrationscripts/20220927_add_snapshot.go:27]: Why not keep it along side "CommitLineChange"? Any particular reasons?

**Comment by klesh** [comment]
> Hello @klesh, I add a new table named snapshot to repo_snapshot. using this table we can archieve some metrics. I also update grafana gitextractor dashboard too. Like: image If this is OK, I will create a new PR to submit the dashboard change.

Good job, the PR seems alright overall except for one minor problem. Please resolve that and then we can merge it. Thanks for your contribution.  😄

File: plugins/gitextractor/models/migrationscripts/20220927_add_snapshot.go (hunk 1 above)
Comment (xgdyp) [reply plugins/gitextractor/models/migrationscripts/20220927_add_snapshot.go:27]: Do you mean why don't keep Snapshot struct in models/domainlayer/code/commit.go? I think this Snapshot belongs to gitextractor plugin and I update gitextractor plugin and create migrationscripts in plugin.

Comment (klesh) [reply plugins/gitextractor/models/migrationscripts/20220927_add_snapshot.go:27]: I don't think so, it should be part of the domain layer, it describes the latest status of the repository, not necessary bound to `gitext` itself.

Comment (xgdyp) [reply plugins/gitextractor/models/migrationscripts/20220927_add_snapshot.go:27]: ok, I'll fix it.

**Comment by xgdyp** [comment]
hello @klesh, I need your approve to run tests.

File: models/domainlayer/code/commit.go [hunk 2]
```diff
@@ -84,3 +84,15 @@ type CommitLineChange struct {
 func (CommitLineChange) TableName() string {
 	return "commit_line_change"
 }
+
+type Snapshot struct {
```
Comment (klesh) [inline models/domainlayer/code/commit.go:88]: Can we name it RepoSnapshot?

Comment (xgdyp) [reply models/domainlayer/code/commit.go:88]: OK, I think `RepoSnapthot` is better

**Review by klesh — APPROVED** [review]
LGTM

**Pull request merged**